engine-precompiles/src/xcc.rs

```diff
@@ -0,0 +1,154 @@
+    pub const ADDRESS: Address =
+        crate::make_address(0x516cded1, 0xd16af10cad47d6d49128e2eb7d27b372);
+}
+
+impl<I: IO> Precompile for CrossContractCall<I> {
+    fn required_gas(_input: &[u8]) -> Result<EthGas, ExitError> {
+        Ok(costs::CROSS_CONTRACT_CALL)
+    }
+
+    fn run(
+        &self,
+        input: &[u8],
+        target_gas: Option<EthGas>,
+        context: &Context,
+        is_static: bool,
+    ) -> EvmPrecompileResult {
+        if let Some(target_gas) = target_gas {
+            if Self::required_gas(input)? > target_gas {
+                return Err(ExitError::OutOfGas);
+            }
+        }
+
+        // It's not allowed to call cross contract call precompile in static or delegate mode
+        if is_static {
+            return Err(ExitError::Other(Cow::from(ERR_STATIC)));
+        } else if context.address != cross_contract_call::ADDRESS.raw() {
+            return Err(ExitError::Other(Cow::from(ERR_DELEGATE)));
+        }
+
+        let sender = context.caller;
+        let target_account_id = create_target_account_id(sender, self.engine_account_id.as_ref());
+        // TODO: Is it ok to use Borsh to read the input? It might not be very friendly to construct the input in Solidity...
+        let args = CrossContractCallArgs::try_from_slice(input)
+            .map_err(|_| ExitError::Other(Cow::from(ERR_INVALID_INPUT)))?;
+        let promise = match args {
+            CrossContractCallArgs::Eager(call) => {
+                let call_gas = match &call {
+                    PromiseArgs::Create(call) => call.attached_gas,
+                    PromiseArgs::Callback(cb) => cb.base.attached_gas + cb.callback.attached_gas,
+                };
+                PromiseCreateArgs {
+                    target_account_id,
+                    method: ROUTER_EXEC_NAME.into(),
+                    args: call
+                        .try_to_vec()
+                        .map_err(|_| ExitError::Other(Cow::from(ERR_SERIALIZE)))?,
+                    attached_balance: ZERO_YOCTO,
+                    attached_gas: costs::ROUTER_EXEC + call_gas,
+                }
+            }
+            CrossContractCallArgs::Delayed(call) => PromiseCreateArgs {
+                target_account_id,
+                method: ROUTER_SCHEDULE_NAME.into(),
+                args: call
+                    .try_to_vec()
+                    .map_err(|_| ExitError::Other(Cow::from(ERR_SERIALIZE)))?,
+                attached_balance: ZERO_YOCTO,
+                // We don't need to add any gas to the amount need for the schedule call
+                // since the promise is not executed right away.
+                attached_gas: costs::ROUTER_SCHEDULE,
+            },
+        };
+
+        let promise_log = Log {
+            address: cross_contract_call::ADDRESS.raw(),
+            topics: Vec::new(),
+            data: promise
+                .try_to_vec()
```

Comment:
Thinking out loud here: it is unfortunate that we need to serialize the promise here, and then deserialize it back later. This could be potentially saved, but it is not trivial (and outside of the scope of this PR).

Reply:
It's not here that we re-serialize, it is when we serialize `call` (in the `promise.args` field). This could be avoided by manually matching on the first byte of the input (which selects between `Eager` and `Delayed` variants), but this would be more error prone because we would need to remember to change the parsing here any time `CrossContractCallArgs` was updated. Another disadvantage of not parsing the input entirely is that we then are not validating it is a real `PromiseArgs` argument, so it would not be until the call hit the router contract that it would fail. I think it is a much better developer experience to have it fail in the EVM.

For these reasons, I think it worth spending the extra gas to do the round-trip serialization.